library/gui/screen/src/main/java/org/quiltmc/qsl/screen/mixin/client/MinecraftClientMixin.java

```diff
@@ -0,0 +1,84 @@
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.gui.screen.Screen;
+
+@Mixin(MinecraftClient.class)
+abstract class MinecraftClientMixin {
+	@Shadow
+	public Screen currentScreen;
+
+	@Unique
+	private Screen quilt$tickingScreen;
+
+	@Inject(method = "setScreen", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/Screen;removed()V", shift = At.Shift.AFTER))
+	private void onScreenRemove(@Nullable Screen screen, CallbackInfo ci) {
+		ScreenEvents.remove(this.currentScreen).invoker().onRemove(this.currentScreen);
+	}
+
+	@Inject(method = "stop", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/Screen;removed()V", shift = At.Shift.AFTER))
+	private void onScreenRemoveBecauseStopping(CallbackInfo ci) {
+		ScreenEvents.remove(this.currentScreen).invoker().onRemove(this.currentScreen);
+	}
+
+	// Synthetic method in `tick`
+	// These two injections should be caught by "Screen#wrapScreenError" if anything fails in an event and then rethrown in the crash report
+	@Inject(method = "method_1572", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/Screen;tick()V"))
+	private void beforeScreenTick(CallbackInfo ci) {
+		// Store the screen in a variable in case someone tries to change the screen during this before tick event.
+		// If someone changes the screen, the after tick event will likely have class cast exceptions or an NPE.
+		this.quilt$tickingScreen = this.currentScreen;
+		ScreenEvents.beforeTick(this.quilt$tickingScreen).invoker().beforeTick(this.quilt$tickingScreen);
+	}
+
+	// Synthetic method in `tick`
+	@Inject(method = "method_1572", at = @At("TAIL"))
+	private void afterScreenTick(CallbackInfo ci) {
+		ScreenEvents.afterTick(this.quilt$tickingScreen).invoker().afterTick(this.quilt$tickingScreen);
+		// Finally set the currently ticking screen to null
+		this.quilt$tickingScreen = null;
+	}
+
+	// The LevelLoadingScreen is the odd screen that isn't ticked by the main tick loop, so we fire events for this screen.
+	// We Coerce the package-private inner class representing the world load action so we don't need an access widener.
+	@Inject(method = "startIntegratedServer(Ljava/lang/String;Ljava/util/function/Function;Ljava/util/function/Function;ZLnet/minecraft/client/MinecraftClient$WorldLoadAction;)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/LevelLoadingScreen;tick()V"))
+	private void beforeLoadingScreenTick(CallbackInfo ci) {
+		// Store the screen in a variable in case someone tries to change the screen during this before tick event.
+		// If someone changes the screen, the after tick event will likely have class cast exceptions or throw a NPE.
+		this.quilt$tickingScreen = this.currentScreen;
+		org.quiltmc.qsl.screen.api.client.ScreenEvents.beforeTick(this.quilt$tickingScreen).invoker().beforeTick(this.quilt$tickingScreen);
```

Comment:
Why is this class fully qualified despite the fact that it is already imported?

Reply:
Oops that must be a leftover from porting will fix

Reply:
Fixed.